### PR TITLE
Add database module

### PR DESCRIPTION
**Pull request by lowlydba**
## Description

Add a database module.

## How Has This Been Tested?
Integration tests! 🎉 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue) - Fixes #
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have read/followed the [**CONTRIBUTING**](https://github.com/LowlyDBA/lowlydba.sqlserver/blob/main/.github/CONTRIBUTING.md) document.
- [x] I have read/followed the [PR Quick Start Guide](https://github.com/ansible/community-docs/blob/main/create_pr_quick_start_guide.rst)  # TODO: update with link to published doc
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 | Module | Description |
 | ------ | ----------- |
+| [database](plugins/modules/database.ps1) | Creates and configures a database. |
```
Comment (briantist) [inline README.md:21]: I promise this'll be unnecessary soon(ish) with the docs build stuff. I mean you could make it unnecessary today by using surge.sh as a host, but I figure you'll be more interested in GitHub pages as a docs host.. I just gotta find the time...

File: plugins/module_utils/_SqlServerUtils.psm1 [hunk 2]
```diff
@@ -35,9 +35,9 @@ function ConvertTo-HashTable {
         foreach ($property in $objectProperty) {
             $propertyName = $property.Name
             switch -Wildcard ($property.TypeNameOfValue) {
-                "Microsoft*Collection" { $outputHash[$propertyName] = [string[]]$Object.$propertyName.Name }
-                "Microsoft*" { $outputHash[$propertyName] = $Object.$propertyName.Name }
-                "SqlCollaborative.DbaTools.Parameter.DbaInstanceParameter" { $outputHash[$propertyName] = $Object.$propertyName.FullName }
+                "Microsoft.*Collection" { $outputHash[$propertyName] = [string[]]$Object.$propertyName.Name; break }
```
Comment (briantist) [inline plugins/module_utils/_SqlServerUtils.psm1:38]: qq why do we need dbatools objects to be converted to hashtables?

File: plugins/modules/database.ps1 [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        recovery_model = @{type = 'str'; required = $false; choices = @('Full', 'Simple', 'BulkLogged') }
+        data_file_path = @{type = 'str'; required = $false }
+        log_file_path = @{type = 'str'; required = $false }
+        owner_name = @{type = 'str'; required = $false; }
+        maxdop = @{type = 'int'; required = $false; }
+        secondary_maxdop = @{type = 'int'; required = $false; }
+        compatibility = @{type = 'str'; required = $false; }
+        rcsi = @{type = 'bool'; required = $false; }
+        state = @{type = 'str'; required = $false; default = 'present'; choices = @('present', 'absent') }
+    }
```
Comment (briantist) [inline plugins/modules/database.ps1:29]:
```suggestion
    }
    required_together = @(
        ,@('sql_username', 'sql_password')
    )
```
These two are required together, no?
Details in : https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general_windows.html

File: plugins/modules/database.ps1 [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            Database = $database
+            EnableException = $true
+            Confirm = $false
+        }
+        Remove-DbaDatabase @dropHash
+        $module.Result.changed = $true
+    }
+    #TODO: Is it wrong to not return a useless dictionary of dropped db
+    # attributes here? It makes different states = different output values.
+    # or maybe default to a few basic properties (see memory module's output)
```
Comment (briantist) [inline plugins/modules/database.ps1:82]: Might be something to consider as diff output for `--diff` mode?

File: plugins/modules/database.ps1 [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                $newDbParams.Add("DataFilePath", $dataFilePath)
+            }
+            if ($null -ne $logFilePath) {
+                $newDbParams.Add("LogFilePath", $logFilePath)
+            }
+            $output = New-DbaDatabase @newDbParams
+            $module.Result.changed = $true
+        }
+        catch {
+            $module.FailJson("Creating database [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:106]:
```suggestion
            $module.FailJson("Creating database [$database] failed.", $_)
```
I believe this second parameter should be the whole exception/errorrecord object.

File: plugins/modules/database.ps1 [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                    }
+                    $null = Set-DbaDbOwner @setDbParams
+                }
+            }
+            # Re-fetch the output since Owner is a read-only property
+            $output = Get-DbaDatabase @getDatabaseHash
+            $module.Result.changed = $true
+        }
+        catch {
+            $module.FailJson("Setting database owner for [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:130]:
```suggestion
            $module.FailJson("Setting database owner for [$database] failed.", $_)
```

File: plugins/modules/database.ps1 [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                        Confirm = $false
+                    }
+                    $null = Set-DbaDbRecoveryModel @recoveryModelHash
+                }
+                $output.RecoveryModel = $recoveryModel
+                $module.Result.changed = $true
+            }
+        }
+        catch {
+            $module.FailJson("Setting recovery model for [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:154]:
```suggestion
            $module.FailJson("Setting recovery model for [$database] failed.", $_)
```

File: plugins/modules/database.ps1 [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                        EnableException = $true
+                    }
+                    $null = Set-DbaDbCompatibility @compatHash
+                }
+                $output.Compatibility = $compatibility
+                $module.Result.changed = $true
+            }
+        }
+        catch {
+            $module.FailJson("Setting Compatibility for [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:178]:
```suggestion
            $module.FailJson("Setting Compatibility for [$database] failed.", $_)
```

File: plugins/modules/database.ps1 [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+                if (-not($checkMode)) {
+                    $server.Databases[$database].IsReadCommittedSnapshotOn = $rcsiEnabled
+                    $server.Databases[$database].Alter()
+                }
+                $output.RCSI = $rcsiEnabled
+                $module.Result.changed = $true
+            }
+        }
+        catch {
+            $module.FailJson("Setting Read Commmitted Snapshot Isolation for [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:197]:
```suggestion
            $module.FailJson("Setting Read Commmitted Snapshot Isolation for [$database] failed.", $_)
```

File: plugins/modules/database.ps1 [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                if (-not($checkMode)) {
+                    $server.Databases[$database].MaxDop = $maxDop
+                    $server.Databases[$database].Alter()
+                }
+                $output.MaxDop = $MaxDOP
+                $module.Result.changed = $true
+            }
+        }
+        catch {
+            $module.FailJson("Setting MAXDOP for [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:218]:
```suggestion
            $module.FailJson("Setting MAXDOP for [$database] failed.", $_)
```

File: plugins/modules/database.ps1 [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+                if (-not($CheckMode)) {
+                    $server.Databases[$database].MaxDopForSecondary = $secondaryMaxDOP
+                    $server.Databases[$database].Alter()
+                }
+                $output.SecondaryMaxDop = $secondaryMaxDop
+                $module.Result.changed = $true
+            }
+        }
+        catch {
+            $module.FailJson("Setting MaxDop for secondary mode for [$database] failed.", $_.Exception.Message)
```
Comment (briantist) [inline plugins/modules/database.ps1:238]:
```suggestion
            $module.FailJson("Setting MaxDop for secondary mode for [$database] failed.", $_)
```

File: plugins/modules/database.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    type: int
+  secondary_maxdop:
+    description:
+      - MAXDOP value for the database when it is a non-primary replica in an availability group.
+    required: false
+    type: int
+  compatibility:
+    description:
+      - Compatibility mode for the database. Follows the format of "Version90", "Version100", and so on.
+        String is validated by Set-DbaDbCompatibility.
```
Comment (briantist) [inline plugins/modules/database.py:55]:
```suggestion
        String is validated by C(Set-DbaDbCompatibility).
```

File: plugins/modules/database.py [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    required: false
+    type: str
+  rcsi:
+    description:
+      - Whether or not to enable Read Committed Snapshot Isolation.
+    required: false
+    type: bool
+  state:
+    description:
+      - Whether or not the database should be created, remain, or be dropped.
```
Comment (briantist) [inline plugins/modules/database.py:65]: kind of a nit but the description is a little confusing when compared to the `choices`. Maybe it should be like:
```suggestion
      - Whether or not the database should be C(present) or C(absent).
```
don't feel terribly strongly about it though.

File: plugins/modules/database.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    required: false
+    type: str
+    default: 'present'
+    choices: ['present', 'absent']
+author: "John McCall (@lowlydba)"
+notes:
+  - Check mode is supported.
```
Comment (briantist) [inline plugins/modules/database.py:72]: I could be wrong but I don't think this needs to be called out manually. Actually maybe that relies on the new "attributes" thing.. https://github.com/ansible-collections/community.hashi_vault/issues/197

File: tests/integration/targets/database/aliases [hunk 15]
```diff
@@ -0,0 +1,2 @@
+context/target
```
Comment (briantist) [inline tests/integration/targets/database/aliases:1]: can be left out (will be detected as a module and known to be in target context) but mentioning more as an informational thing, can leave it

**Review by briantist — COMMENTED** [review]
Overall looks good to me, mostly small comments.

Some additional testing of mor parameters and combinations might be nice, and using `win_shell` or `win_powershell` tasks to directly (outside of the module) check the state of the database to verify functionality might be good to shore up the tests too, but doesn't have to be done in this PR.

**Comment by lowlydba** [comment]
> Overall looks good to me, mostly small comments.
> 
> Some additional testing of mor parameters and combinations might be nice, and using `win_shell` or `win_powershell` tasks to directly (outside of the module) check the state of the database to verify functionality might be good to shore up the tests too, but doesn't have to be done in this PR.

Agreed, will probably start hacking through more robust integration tests / windows stuff once I get through the backlog of converting existing modules.

File: plugins/module_utils/_SqlServerUtils.psm1 (hunk 2 above)
Comment (briantist) [reply plugins/module_utils/_SqlServerUtils.psm1:38]: not blocking but still curious about this

**Review by briantist — APPROVED** [review]
When there are multiple suggested edits, you can batch them by looking at them in the Files view (Files changed); there you have the option to commit directly or to "Add to batch" and then you can do a single commit with several suggestions. It took me _way_ too long to realize this myself...

**Pull request merged**